Advanced item level versioning for DS 7

parphis opened the pull request:
## References
* Fixes https://jira.lyrasis.org/browse/DS-4586
* Fixes #7919 

## Description
This PR is the modification of the #3376 one in order to apply it for the version 7.x

As per the linked ticket
> We were thinking of improving item versioning. Our aim was to:
>
> 1. provide a persistent Handle for each item version
> 2. provide a persistent Handle that references always the most recent version

> While the current DSpace Handle versioning scheme allows (1), it does not provide (2). Basically our idea is the scheme of how  arXiv.org provides persistent item references.

## Instructions for Reviewers
With these modifications we may add a configuration property to the basic config: `initialVersionSuffix`.
> If this option is missing or if it has an empty value then the item versioning acts like before. However, if it has a string of ".1" then the Handle of the very first version of the item will get a version suffix.

Most of the modifications have been applied on the `HandleServiceImpl.java` in order to add a new parameter for the function `findHandleInternal`. With the new boolean argument we can make difference between the calls:
 1. Are we trying to resolve a handle - use true.
 2. Are we checking the existence of a handle during id creation - use false.

We also has changed the `createId` function to use the `initialVersionSuffix` and to determine wheter the new handle id will be used for an item or any other resource?

A new function `findByVersion` within `HandleDAO` class aims the modified functionality in the `findHandleInternal`. It returns a handle list within the Version table based on the given handle.

Applying our modifications result the followings when using the item versioning feature (assuming the `initialVersionSuffix` is set to _.1_):
 1. Every new item gets a versioned handle like 123456789/1.1
 2. Every new version added will get new version numbers increased by one.
 3. Linking to a handle with a specific version number renderers the exact item. However linking to an item without any version suffix (like 123456789/1) returns the most recent version of the item.

In case of missing option `initialVersionSuffix` or leaving it empty means the original item versionig will be used.

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

parphis committed:
DSpace/DSpace/pull/3376 for DS 7

tdonohue commented:
@abollini : as your team is working on Item Level versioning support for 7.1 (in #3371), I'd appreciate your feedback on this suggestion/improvement on how version numbers are assigned to items.  So, I'm adding you as a reviewer (feel free to delegate to another on your team if needed).  This may not necessarily need to be in 7.1, but I wanted to alert you to this work in case you found it to be a worthwhile "quick fix".

abollini reviewed (changes requested):
Sorry to come back to this PR so late. I found an incomplete review that I have never published.
At the current stage this PR requires some update to be considered
- we need additional javadoc on the service methods to understand the differences between the `findByVersion` and `findByPrefix` methods as they seem to be duplicate
- the `INITIAL_VERSION_SUFFIX` cannot be really anything else than .1 or the empty string so it should be better eventually to have a boolean attribute that switch from the current behavior to the new one
- test must be provided to proof that the new configuration works as expected

parphis committed:
Merge branch 'main' into advanced-item-level-versioning-7_x

parphis committed:
Merge branch 'main' of github.com:bme-omikk/DSpace into advanced-item-level-versioning-7_x

parphis committed:
Added integration test; removed initial_version_suffix var.

tdonohue commented:
@parphis : FYI, It looks like your most recent commit (to add integration tests) is failing tests

parphis commented:
@tdonohue Thanks, I'm working on it continuosly locally.

parphis committed:
Modifications based on suggestions from  @abollini

parphis commented:
@abollini I changed the code as per you requests/suggestions:

 - I create integration tests to see whether the handle creation and resolving processes are working well.
 - I removed the constant and changed the code based on this modification.
 - I also removed the `findByVersion` function and changed the `findByHandle` one in order to better express changes.

Please let me know if you find other problems. I have a bit of experience in Java and integration testing methods but I am trying to do the best.

@tdonohue You wrote in an e-mail in the last weeks that we should notice you until 1th of August if we need to add new features for the version 7.4. So I would like to ask you to add this PR to the version 7.4. I try to do my best to correct any errors before the release date of October. Thank you.

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+                          .withSubject("ExtraEntry")
+                          .build();
+
+        System.out.println("Item handle is: ");
+        System.out.println(item.getHandle());
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+                        .andExpect(content().contentType(contentType))
+                        .andExpect(jsonPath("$.handle", endsWith(".1")));
+
+        getClient(token).perform(get("/api/versioning/versions/" + version1st.getID()))
+            .andExpect(status().isOk())
+            .andExpect(jsonPath("$", Matchers.is(VersionMatcher.matchEntry(version1st))))
+            .andExpect(jsonPath("$._links.versionhistory.href", Matchers.allOf(Matchers.containsString(
+                                "api/versioning/versions/" + version1st.getID() + "/versionhistory"))))
+            .andExpect(jsonPath("$._links.item.href", Matchers.allOf(Matchers.containsString(
+                                "api/versioning/versions/" + version1st.getID() + "/item"))))
+            .andExpect(jsonPath("$._links.self.href", Matchers.allOf(Matchers.containsString(
+                                "api/versioning/versions/" + version1st.getID()))));
+
+        getClient(token).perform(get("/api/versioning/versions/" + version2nd.getID()))
+            .andExpect(status().isOk())
+            .andExpect(jsonPath("$", Matchers.is(VersionMatcher.matchEntry(version2nd))))
+            .andExpect(jsonPath("$._links.versionhistory.href", Matchers.allOf(Matchers.containsString(
+                                "api/versioning/versions/" + version2nd.getID() + "/versionhistory"))))
+            .andExpect(jsonPath("$._links.item.href", Matchers.allOf(Matchers.containsString(
+                                "api/versioning/versions/" + version2nd.getID() + "/item"))))
+            .andExpect(jsonPath("$._links.self.href", Matchers.allOf(Matchers.containsString(
+                                "api/versioning/versions/" + version2nd.getID()))));
```
you need to check that the new versions will get .2 and .3

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+        String resolvedVersionHandle = HandleServiceFactory.getInstance().getHandleService().resolveToURL(
+                                          context, version1st.getItem().getHandle());
+
+        System.out.println("resolved handle with suffix: " + resolvedHandleWithSuffix);
+        System.out.println("resolved handle without suffix: " + resolvedHandleWithoutSuffix);
+
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+        resolvedHandleWithoutSuffix = HandleServiceFactory.getInstance().getHandleService().resolveToURL(
+                                                context, item.getHandle().split("\\.")[0]);
+
+        System.out.println("resolved handle with suffix: " + resolvedHandleWithSuffix);
+        System.out.println("resolved handle without suffix: " + resolvedHandleWithoutSuffix);
+
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+
+    @Test
+    public void resolveToURLWithFirstSuffixTest() throws Exception {
+        System.out.println("resolveToURLTest");
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+     */
+    @Test
+    public void itemWithFirstVersionSuffixTest() throws Exception {
+        System.out.println("itemWithVersionSuffixTest");
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+     */
+    @Test
+    public void itemWithoutVersionSuffixTest() throws Exception {
+        System.out.println("itemWithoutVersionSuffixTest");
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+            context, item, "First version of the item without .1 suffix")
+                .build();
+
+        System.out.println("hello from resolveToURLTest: " + item.getHandle());
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+
+    @Test
+    public void resolveToURLWithoutFirstSuffixTest() throws Exception {
+        System.out.println("resolveToURLTest");
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+                          .withSubject("ExtraEntry")
+                          .build();
+
+        System.out.println("hello from resolveToURLTest: " + item.getHandle());
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+        String resolvedHandleWithoutSuffix = HandleServiceFactory.getInstance().getHandleService().resolveToURL(
+                                                context, item.getHandle().split("\\.")[0]);
+
+        System.out.println("resolved handle with suffix: " + resolvedHandleWithSuffix);
+        System.out.println("resolved handle without suffix: " + resolvedHandleWithoutSuffix);
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+        resolvedHandleWithoutSuffix = HandleServiceFactory.getInstance().getHandleService().resolveToURL(
+                                                context, item.getHandle().split("\\.")[0]);
+
+        System.out.println("resolved handle with suffix: " + resolvedHandleWithSuffix);
+        System.out.println("resolved handle without suffix: " + resolvedHandleWithoutSuffix);
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+                          .withSubject("ExtraEntry")
+                          .build();
+
+        System.out.println("hello from resolveUrlToHandleTest: " + item.getHandle());
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+
+    @Test
+    public void resolveUrlToHandleTest() throws Exception, SQLException {
+        System.out.println("resolveUrlToHandleTest");
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+        String resolvedHandleWithoutSuffix = HandleServiceFactory.getInstance().getHandleService().resolveUrlToHandle(
+                                                context, baseUrl + item.getHandle().split("\\.")[0]);
+
+        System.out.println("resolved handle with suffix: " + resolvedHandleWithSuffix);
+        System.out.println("resolved handle without suffix: " + resolvedHandleWithoutSuffix);
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+
+    @Test
+    public void resolveToObjectTest() throws Exception, SQLException {
+        System.out.println("resolveToObjectTest");
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+                          .withSubject("ExtraEntry")
+                          .build();
+
+        System.out.println("hello from resolveToObjectTest: " + item.getHandle());
```
please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+    }
+
+    @Test
+    public void resolveToURLWithoutFirstSuffixTest() throws Exception {
```
you should perform a test also with two versions to verify that the .1 version resolve correctly and the most recent handle resolve correctly too.

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+    }
+
+    @Test
+    public void resolveUrlToHandleTest() throws Exception, SQLException {
```
you should perform a test also with two versions to verify that the .1 version resolve correctly and the most recent handle resolve correctly too.

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/AdvancedItemLevelVersioningIT.java:
```diff
@@ -0,0 +1,384 @@
+        DSpaceObject resolvedDsoWithoutSuffix = HandleServiceFactory.getInstance().getHandleService().resolveToObject(
+                                                context, item.getHandle().split("\\.")[0], true);
+
+        System.out.println("resolved dso with suffix: " + resolvedDsoWithSuffix.getHandle());
+        System.out.println("resolved dso without suffix: " + resolvedDsoWithoutSuffix.getHandle());
```
please remove

abollini reviewed (changes requested):
Hi @parphis thanks for your work here. I have made some comment inline in the code but just before to submit my review I have noted that we already have in the codebase an alternative implementation of the VersionedHandleProvider that looks to implement exactly the same scenario that you want to achieve. Moreover, this implementation follow the main suggestion that I have put in the review of your code, the "most recent version handle" is stored in the database.
Can you take a look to `org.dspace.identifier.VersionedHandleIdentifierProviderWithCanonicalHandles` to see if I'm right? if this is the case we should keep the IT that you have introduced in this PR and verify them against this implementation

tdonohue commented:
@parphis : It looks like this PR is stalled.  Are you still planning to work on this for our upcoming 7.4 release (due in early Oct)? If so, we'd need updates to this in the next few weeks to help move it forward.  If not, we can also reschedule for 7.5 (due in Feb)

parphis commented:
@tdonohue as per the notes from @abollini I have tested the versioning with the recommended settings (with `VersionedHandleIdentifierProviderWithCanonicalHandles` uncommented) last week and it has been determined that it fulfills our requirements for the versioning of the handles. This means that changes I have made on the code can be dropped.

Currently, I am fine tuning the created integration test in order to apply the suggestions of @abollini 

Please let me know if I should open a new PR for this new integration test or leave it here? Thank you!

parphis commented on dspace-api/src/main/java/org/dspace/handle/dao/impl/HandleDAOImpl.java:
```diff
@@ -63,20 +63,46 @@ public List<Handle> getHandlesByDSpaceObject(Context context, DSpaceObject dso)
+    public Handle findByHandle(Context context, String handle, boolean getMostRecentVersion) throws SQLException {
+        Query query;
+
+        if (getMostRecentVersion) {
+            query = createQuery(context,
+                "SELECT h " +
+                "FROM Handle h, DSpaceObject dso, Version v " +
+                "WHERE h.dso = dso.id " +
+                "and v.item = dso.id " +
+                "and h.handle like :handle " +
+                "ORDER BY v.versionDate desc "
+                );
+
+            query.setParameter("handle", handle + "%");
+        } else {
+            query = createQuery(context,
+                "SELECT h " +
+                "FROM Handle h " +
+                "LEFT JOIN FETCH h.dso " +
+                "WHERE h.handle = :handle ");
```
@abollini As I described below the `VersionedHandleIdentifierProviderWithCanonicalHandles` option does what we need regarding the versioning. So I close this PR and will open a new one with the modified integration tests. Thank you for your support.

parphis commented:
>  if this is the case we should keep the IT that you have introduced in this PR and verify them against this implementation

@abollini @tdonohue 
I would like to run an integration test to see if the handle versioning is working well using the `org.dspace.identifier.VersionedHandleIdentifierProviderWithCanonicalHandles`.

I copy the `identifier-service.xml` into the folders `dspace-api/src/test/data/dspaceFolder/config/spring/api` and `dspace-server-webapp/src/test/data/dspaceFolder/config/spring/api` with the uncommented lines
```XML
<bean id="org.dspace.identifier.HandleIdentifierProvider" class="org.dspace.identifier.VersionedHandleIdentifierProviderWithCanonicalHandles" scope="singleton">
        <property name="configurationService" ref="org.dspace.services.ConfigurationService"/>
    </bean>
```
I checked the file `dspace-server-webapp/target/testing/dspace/config/spring/api/identifier-service.xml` and found that the uncommented lines are there.
But when I run the integration test and print out the generated handles I see the default versioning functionality. That is, the origin item gets a handle without a version suffix, the second version gets a .2, the third gets a .3 etc. So it seems that the test does not use the `VersionedHandleIdentifierProviderWithCanonicalHandles` class as expected.
Could you please support me on how should I set up the IT in order to use the canonical handles version? Is there any function in the test environment to modify this configuration at load time (like the configuration service)? (Sorry I am not using the Java/Spring tests regularly so I have limited knowledge on it). Thanks!
